docs/arcadia-users-group/20231104-testing-concepts/lesson.md

```diff
@@ -0,0 +1,572 @@
+
+The major benefit a testing framework provides is exactly that, a utility to find and run the tests automatically.
+With pytest, this is the command-line tool called `pytest`.
+When `pytest` is run, it will search all directories below where it was called, find all of the Python files in these directories whose names start or end with `test`, import them, and run all of the functions and classes whose names start with `test` or `Test`.
+This automatic registration of test code saves tons of time and provides a consistent organization framework across Python projects.
+
+When you run `pytest`, it will print a dot (`.`) on the screen for every test that passes, an `F` for every test that fails or where there was an unexpected error.
+In rarer situations you may also see an `s` indicating a skipped tests (because the test is not applicable on your system) or a `x` for a known failure (because the developers could not fix it promptly).
+After the dots, pytest will print summary information.
+
+**Challenge 1: Altering functions to pass all tests**
+
+Without changing the tests, alter the mean.py file from the previous section until it passes.
+When it passes, `pytest` will produce results like the following:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+========================== 5 passed in 2.68 seconds ===========================
+```
+
+<details>
+ <summary> Challenge solution </summary>
+
+There are many ways this challenge could be solved.
+One way is to check for the presence of complex numbers before calculating the mean.
+
+```
+def mean(num_list):
+    if any(isinstance(num, complex) for num in num_list):
+       return NotImplemented
+    else:
+        try:
+            return sum(num_list)/len(num_list)
+        except ZeroDivisionError :
+            return 0
+        except TypeError as detail :
+            msg = "The algebraic mean of an non-numerical list is undefined.\
+                   Please provide a list of numbers."
+            raise TypeError(detail.__str__() + "\n" +  msg)
+```
+
+</details>
+
+
+Note, using `pytest -v` will result in `pytest` listing which tests are executed and whether they pass or not:
+
+```
+pytest
+```
+
+```
+collected 5 items
+
+test_mean.py .....
+
+test_mean.py::test_ints PASSED
+test_mean.py::test_zero PASSED
+test_mean.py::test_double PASSED
+test_mean.py::test_long PASSED
+test_mean.py::test_complex PASSED
+
+========================== 5 passed in 2.57 seconds ===========================
+```
+
+As we write more code, we would write more tests, and pytest would produce more dots.
+
+## Integration tests
+
+Integration tests focus on gluing code together or the results of code when multiple functions are used.
+See below for an conceptual example of an integration test.
+
+Consider three functions `a()`, `b()`, and `c()` as a simplistic example.
+Function `a()` increments a number by one, `b()` multiplies a number by two, and `c()` composes them as defined below:
```

Comment:
another nit: I would favor renaming these functions with more explicit names both for clarity and to avoid implicitly sanctioning the use of ambiguous/meaningless single-letter names

Reply:
oh i totally agree, good catch thank you!!